### ISSUE TITLE
Lap 2 and subsequent laps start a minute early

### ISSUE THREAD

**Issue by ntadmin** [GoldenCheetah/GoldenCheetah#3457]
I recently returned to using golden cheetah to look at some data, which is imported from Strava. The data I was looking at had four laps "laps", which are correctly described in the summary on Golden Cheetah:

Lap 1 duration 22:52
Lap 2 duration 10:06 (by inference starts 22:52)
Lap 3 duration 20:01 (by inference starts 32:58)
Lap 4 duration 14:04 (by inference starts 42:59)

This matches the data on Strava. All the imported data (power, HR, etc ) seems correct. For this piece, the start of lap 2 and 4 are especially evident  from the power data.

However, when the data is viewed in, say "Ride" (or any there view). Lap 2 starts at 21:51, lap 3 at 31:58, lap 4 at 41:59, ie they all (apart from lap 1) start a minute early. This shows it pretty clearly:
![Screenshot 2020-05-23 at 12 16 40](https://user-images.githubusercontent.com/8012540/82729409-4495b480-9cef-11ea-9198-909d31019a52.png)

And the original data on strava, not super clear but the start of lap 2 and end of lap 4 bars at the top can be seen to correctly align with the start and end of the "higher power" section:
![Screenshot 2020-05-23 at 12 20 59](https://user-images.githubusercontent.com/8012540/82729486-def5f800-9cef-11ea-8fe8-df0e5ebee7e0.png)

I see the same effect on other rides.

Apologies if this is known or listed elsewhere. I had a look here and on the forum and couldn't find an issue or discussion with "lap" in the title.

**Comment by ntadmin** [comment]
Sorry, should have put version information above, here it is:
![Screenshot 2020-05-23 at 12 23 49](https://user-images.githubusercontent.com/8012540/82729541-44e27f80-9cf0-11ea-90dc-0a8c9d6fe37b.png)

**Comment by amtriathlon** [comment]
There was a previous report similar to this: https://github.com/GoldenCheetah/GoldenCheetah/issues/3254
If you attach the original (FIT) file we could upload it to Strava and sync to try to recreate the problem.

**Comment by ntadmin** [comment]
I think that this is the file, taken directly from the Garmin - it has the right date in the title! The processing route was standard as I understand it: Device (Garmin Edge 510) -> phone (iPhone old SE)-> Garmin Connect app -> onwards.

[2020-05-23-09-31-39.fit.zip](https://github.com/GoldenCheetah/GoldenCheetah/files/4675977/2020-05-23-09-31-39.fit.zip)

The file itself is the original .fit file and compressed to .zip on a Mac, then uploaded to here.

**Comment by amtriathlon** [comment]
Manually uploading that fit file to Strava and Sync to GC I can not recreate this issue:

![StravaLapAlignment](https://user-images.githubusercontent.com/1444784/82819563-f3b0c800-9e76-11ea-896d-27ab0c1ad2f7.png)

Some questions to try to understand the issue:
This behavior happens on all your activities synced from Strava or only some of them? Is there any pattern?
Is this a permanent issue? I mean, if you delete the local copy and re-sync, the problem happens again?
If you remove the original activity from Strava, manually upload the fit file to Strava and sync, can you recreate the issue?

**Comment by ntadmin** [comment]
Bizarre - the graphs you have posted are what I would expect to see!

So I've done a quick review of some other recent activities. All of the report correct laps on the Summary page (as far as I can remember. About 50% have the laps displayed correctly, about 50% have and error.

So I did a bit more digging with the same ride as before. Now here's a weird thing: when you highlight the lap by moving the cursor into the lap, the highlight is correct (for the first lap) but the lap mark isn't:

![Screenshot 2020-05-25 at 20 16 27](https://user-images.githubusercontent.com/8012540/82838607-9fccd000-9ec4-11ea-81bd-82a4fdc29a98.png)

The second lap has the highlight aligned with the markers:

![Screenshot 2020-05-25 at 20 17 38](https://user-images.githubusercontent.com/8012540/82838659-c8ed6080-9ec4-11ea-8a01-1421cd63c88f.png)

In fact, the bottom of the screen has the laps overlapping (lap 2 is the cyan bar):

![Screenshot 2020-05-25 at 20 18 35](https://user-images.githubusercontent.com/8012540/82838712-f0442d80-9ec4-11ea-932f-ca8293cf1a72.png)

Back to your questions:

1. I deleted the activity and resynced, and the problem was still there.

2. With great fear and trembling I deleted the activity, and re-uploaded it (fit file direct to starve). Exactly the same outcome, including the overlap of laps 1 and 2).

**Comment by amtriathlon** [comment]
Laps in GC are defined by start and end time, so they can overlap, using Edit Interval on the left side bar you can see, and change, interval start and stop time.
We would need to enable logging in the Strava sync to understand why this is happening, but this is a build time option.

**Comment by amtriathlon** [comment]
The following link allows to download a Windows GC v3.6 DEV version with Strava debug enabled:
https://ci.appveyor.com/project/Joern-R/goldencheetah-knhd8/builds/33110629/artifacts
For every activity synced from Strava the debug console will show the responses received, since they are in JSON format the Intervals are easy to identify to check their start and stop times.

**Comment by ntadmin** [comment]
The overlapping laps make a lot of sense, it hadn't occurred to me. So that it consistent with your theory of data import error somehow. I am very, very, happy to download a debug version and see what happens, but awkwardly I us a Mac, and so I can't download the windows version you have so kindly prepared. Two options:
1. How easy is it set up a dev version on my Mac (I have been involved in fairly serious coding and computer stuff in the past)
- or -
2. Are you able to produce a debug version of 3.6 with that flag enabled for Mac?
Thanks

**Comment by ntadmin** [comment]
[As an aside, there is an error in the Mac install guide. The instructions for installing Qt assume that the default version (a) is 4 and (b) will be installed with the default options. Neither is now true. I'm trying again with the macport option.]

**Comment by ntadmin** [comment]
[And if you go down the macport route, qmake is hidden in /opt/local/libexec/qt4/bin/qmake ]

**Comment by amtriathlon** [comment]
The PR includes macOS and Linux builds, currently we can’t access them because transfer.sh was unavailable at build time to upload the results, likely a temporary error, I will retry and update when the binaries are ready. Better not to mix how to build questions with this issue.

**Comment by amtriathlon** [comment]
The PR includes macOS and Linux builds, currently we can’t access them because transfer.sh was unavailable at build time to upload the results, likely a temporary error, I will retry and update when the binaries are ready. Better not to mix how to build questions with this issue.

**Comment by ntadmin** [comment]
[do you want me to report those (and the latest one) "how to build issues" on a separate issue. -would that be helpful or simply noise?]

**Comment by amtriathlon** [comment]
We know the build instructions are outdated, we have a developer forum to assist developers, we have setup CI to automate the process and to avoid the non-developers need to build by themselves. The steps we are using are in the scripts under the Travis folder.

**Comment by ntadmin** [comment]
> We know the build instructions are outdated, we have a developers forum to assist developers, we have setup CI to automate the process and to avoid the non-developers need to build by themselves. The steps we are using are in the scripts under the travis/osx folder.

OK, Thanks, I'll wait for a mac build from you.

**Comment by amtriathlon** [comment]
macOS build is ready and the link can be found at the bottom of the logs for convenience it is copied here: https://transfer.sh/Ax7Xi/GoldenCheetah_v3.6-DEV_x64.dmg, it is supposedly valid for 2 weeks.

**Comment by ntadmin** [comment]
Sequence I followed:
1. Downloaded and installed the 3.6-DEV version of GoldenCheetah
2. Signed in a "Neil", my only profile.
3. viewed the existing data - it was exactly as before with the lap issue on the test case (as one would expect, just a sanity check). It also had the lap issue on the latest data, a ride from this morning.
4. Deleted this entry we are using as a test (23rd May ride, as above).
5. Reauthorised Strava for this version of the application. That failed:

> Step 5 failed: Error retrieving access token, Host requires authentication (204)

(That's the detailed report)

This is odd - I've never had that before.

So, two questions:

1. How do we resolve the Strava authentication issue.
2. Once that is done, where will I find the log file reporting the import that I am assuming you wish to view. I've put this version of Golden Cheetah on my desktop to keep it away from the official versions, currently residing in the standard Applications folder.

**Comment by amtriathlon** [comment]
Sorry, but this attempt will not work, I forgot the secrets required to access online services are not included when building pull requests.

**Comment by ntadmin** [comment]
No worries - I am very happy to build my own copy with the debug flag enabled and create the logging information myselff, where should I look for making that work. I got pretty close - I think I have an issue with MapWindow ofr something like that.

**Comment by amtriathlon** [comment]
To enable Strava access on your own build you need to register with Strava as explained here: https://github.com/GoldenCheetah/GoldenCheetah/wiki/Strava-configuration
I would suggest to use dev forum to search and ask questions about building: https://groups.google.com/forum/m/#!forum/golden-cheetah-developers

**Comment by ntadmin** [comment]
OK, fully built (well, it complains about sqlite3, but it seems to ignore that) and it loads and runs fine. The "test" ride was already missing from the data, and so I then authenticated with Strava successfully. I then synchronised and the new ride added in, with the error on the laps as expected. So, the million dollar question: where will I find the debug log file?

**Comment by amtriathlon** [comment]
Great!
goldencheetah.log is located parallel to your athlete's folder
But, for debugging, I prefer run GC from terminal:
cd /Applications/GoldenCheetah.app/Contents/MacOS
./GoldenCheetah --debug

**Comment by ntadmin** [comment]
OK. log of run:

%./GoldenCheetah.app --debug >> log

1. Start up.
2. Delete activity from May 23rd
3. Strava synchronsise
4. Select the run for the 23rd May to Synchronise
5. Synchronise
6. Check it is there (it is) and shows the lap error (it does)
7. quit.

Log file:
[log.zip](https://github.com/GoldenCheetah/GoldenCheetah/files/4691544/log.zip)

**Comment by ntadmin** [comment]
One other things that may help: when I reviewed the data a bit more closely, it looked more like the discrepancy was 100 seconds, rather than a minute, but I haven’t examined it in detail.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by amtriathlon** [comment]
The relevant parts of the log are the startime:
`"start_date_local":"2020-05-23T09:31:39Z"
`
and laps array:
```
"laps":[
{"id":11634793688, "resource_state":2, "name":"Lap 1", "activity":{"id":3511091699, "resource_state":1}, "athlete":{"id":17553093, "resource_state":1}, "elapsed_time":1371, "moving_time":1371, "start_date":"2020-05-23T08:31:39Z", "start_date_local":"2020-05-23T09:31:39Z", "distance":11533.9, "start_index":0, "end_index":1293, "total_elevation_gain":21.0, "average_speed":8.41, "max_speed":12.6, "average_cadence":96.2, "device_watts":true, "average_watts":169.1, "average_heartrate":121.5, "max_heartrate":153.0, "lap_index":1, "split":1},
{"id":11634793691, "resource_state":2, "name":"Lap 2", "activity":{"id":3511091699, "resource_state":1}, "athlete":{"id":17553093, "resource_state":1}, "elapsed_time":605, "moving_time":605, "start_date":"2020-05-23T08:54:32Z", "start_date_local":"2020-05-23T09:54:32Z", "distance":4693.65, "start_index":1294, "end_index":1876, "total_elevation_gain":25.0, "average_speed":7.76, "max_speed":9.8, "average_cadence":97.6, "device_watts":true, "average_watts":274.7, "average_heartrate":153.5, "max_heartrate":160.0, "lap_index":2, "split":2},
{"id":11634793694, "resource_state":2, "name":"Lap 3", "activity":{"id":3511091699, "resource_state":1}, "athlete":{"id":17553093, "resource_state":1}, "elapsed_time":1200, "moving_time":1200, "start_date":"2020-05-23T09:04:37Z", "start_date_local":"2020-05-23T10:04:37Z", "distance":9620.98, "start_index":1877, "end_index":3062, "total_elevation_gain":41.0, "average_speed":8.02, "max_speed":12.0, "average_cadence":96.9, "device_watts":true, "average_watts":261.0, "average_heartrate":166.0, "max_heartrate":174.0, "lap_index":3, "split":3},
{"id":11634793697, "resource_state":2, "name":"Lap 4", "activity":{"id":3511091699, "resource_state":1}, "athlete":{"id":17553093, "resource_state":1}, "elapsed_time":843, "moving_time":843, "start_date":"2020-05-23T09:24:38Z", "start_date_local":"2020-05-23T10:24:38Z", "distance":4747.0, "start_index":3063, "end_index":3851, "total_elevation_gain":53.0, "average_speed":5.63, "max_speed":9.2, "average_cadence":80.8, "device_watts":true, "average_watts":116.9, "average_heartrate":126.1, "max_heartrate":174.0, "lap_index":4, "split":4}
]
```
We are using start/end index to retrieve the corresponding seconds to mark start/end of each lap, and these don't match in this case for reasons I don't understand.

**Comment by amtriathlon** [comment]
This block:
https://github.com/GoldenCheetah/GoldenCheetah/blob/f18b5470eca7ebb3a1316d897fbde4071a62380e/src/Cloud/Strava.cpp#L956-L960

Can be changed by:
1) Using last_lap as start:
`double start = last_lap; `

2) Using start_date_local to compute start by difference with startime:
`double start = starttime.secsTo(QDateTime::fromString(lap["start_date_local"].toString(), Qt::ISODate);`

The first option assumes laps start at 0 and are contiguous, which is typical of device laps, the second only that start_date_local is correct, which I have not verified in this case, but you can test.

**Comment by ntadmin** [comment]
Thank you - I will work on it, hopefully today.

As an aside, I am intrigued as to why `_local` times are used for this calculation. I can’t see the gain, and if someone rode over a clock change night then things would potentially get messed up? But I may have missed something there. Anyway, not the main point, I’ll get back you you on that one.


<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by ntadmin** [comment]
I was about to have a go with your solutions when I came across another issue: there is also a problem with the final lap - it ends early. Whether you calculate it by looking up the end_index for the lap (104 seconds early) or be adding elapsed time to the time for the start index (59 seconds early). Oddly, there are 3950 data points, but the last lap ends on data point 3851.

So, what seems to be trustworthy are lap start index values (and the times they correspond to):
- elapsed times aren't (they add up correct but are incorrectly distributed)
- times aren't (they seem to be derived from the elapsed times)

So, here is a suggestion for the solution:

- Laps start at the time of the data point given but eh start index, as at present.
- Laps (except the last one) end at the time of the start of the next lap (this makes the same assumption you do, but fixes lap ends to what seem to be "working" points: the starts of laps.
- The final lap ends at the end of the ride

I'm happy to (try and) implement this if you are happy with it. I'm very aware that I am looking at a very small bit of a large project, and I don't know if this approach would be bad for some reason elsewhere in the project.

**Comment by amtriathlon** [comment]
> So, here is a suggestion for the solution:
> 
> * Laps start at the time of the data point given but eh start index, as at present.

Sorry but I don’t follow you, lap start times were the original issue and, at first sight lap duration seems to be ok.
As a guideline we need to look for a general and, hopefully, simple solution, not a fix too specific for this particular case that can fail in others.

>As an aside, I am intrigued as to why _local times are used for this calculation. I can’t see the gain, and if someone rode over a clock change night then things would potentially get messed up?

QTIme handles this issues when you compute time differences with secsTo.

**Comment by ntadmin** [comment]
You are gift I got confused. Let me look again.

**Comment by amtriathlon** [comment]
In the following table Start is just the start_date_local reported minus startime (2nd column) and Stop is Start+elapsed time:
```
name | elapsed_time | 2020-05-23 09:31:39 | Start | Stop
Lap 1 | 1371 | 2020-05-23 09:31:39 | 00:00:00 | 00:22:51
Lap 2 | 605 | 2020-05-23 09:54:32 | 00:22:53 | 00:32:58
Lap 3 | 1200 | 2020-05-23 10:04:37 | 00:32:58 | 00:52:58
Lap 4 | 843 | 2020-05-23 10:24:38 | 00:52:59 | 01:07:02

```
They match the start/stop times we are computing when we import the original FIT file and, since this is the simpler solution without additional assumptions, I think it is the right one.

**Comment by ntadmin** [comment]
Sorry for that confusion - I printed out all the numbers, put them in a spreadsheet, then got something in the wrong order (cause vs. effect) etc tec. Sorry. So:

1. I have changed the code (on my computer) to use the second of your suggestions (the time difference between the "start of lap" time and the "start of ride" time. That seems pretty robust for me, is solid in that no one lap depends on another lap for that to be correct.
2. Using "elapsed time" to get the end time seems correct. So, I have simplified the code in the loop to the following:

```
            // laps?
            if (!each["laps"].isNull()) {
                QJsonArray laps = each["laps"].toArray();
                foreach (QJsonValue value, laps) {
                    QJsonObject lap = value.toObject();

                    double start = starttime.secsTo(QDateTime::fromString(lap["start_date_local"].toString(), Qt::ISODate));
                    double end   = start + lap["elapsed_time"].toDouble();

                    ride->addInterval(RideFileInterval::USER, start, end, lap["name"].toString());
                }

                // Fix distance from laps and fill gaps for pool swims
                fixLapSwim(ride, laps);
            }

```

As you suggested, that seems to work correctly visually on the screen as well.

**Comment by amtriathlon** [comment]
Right, we are in the same page, last_lap was an attempt to solve a case were start index was zero for all laps and it was wrong anyway, current solution doesn't need them so it is ok to remove it.
I will apply this change to master, thanks for your collaboration!

**Comment by ntadmin** [comment]
Thanks - always happy to help, sorry I got confused along the way. I will do a quick review of the other known non correct imports I have over the next few days.

**Comment by ntadmin** [comment]
This morning, on the modded v3.6 version, I deleted every activity that had the overlapping lap 1 and lap 2 (I cannot recall the details of each ride, but that seems da good indicator that something was wrong) for the last 2-3 months and re synced them. They all now have laps which do not over lap and look where it makes sense to me I would have put them! So I'm pretty confident this change is robust, at least for my data!

**Comment by amtriathlon** [comment]
Perfect! I did more testing and found lap swim files still can produce overlapping, which shouldn't be possible by the start/stop mechanic of lap swims, so preserved last_lap and used it to prevent this happening, this doesn't affect the files you send me for testing since start_time_local is ok for them.

**Comment by ntadmin** [comment]
This sounds very odd a question to ask, but did this code get included in the latest dev build (Sep 15 2020; build 4002)? I just updated to it, and the issue is still present. But the code in GitHub is the corrected code I would expect to see, so I am wondering of the older code accidentally was used in the compilation of the executable for the dev version?

**Comment by amtriathlon** [comment]
The source code used for each release is attached together with the binaries, so you can check.
Snapshot builds have a companion txt file listing the last commit included in current master.

**Comment by ntadmin** [comment]
OK, checked the source code from https://github.com/GoldenCheetah/GoldenCheetah/releases/tag/v3.6-DEV2006 and it contains the corrected code, as per our discussion above. However, the file imports are having the same issue as described before we made these changes. I will have a think about how this might be happening, as the code should simply not allow lap 2 to start before lap 1 has ended. Bizarre.
